### PR TITLE
Enable loading full optimizer checkpoints with FSDP

### PR DESCRIPTION
**Pull request by leng-yue**
## What does this PR do?

Making restoring optimizer states from full checkpoint possible, which is a supplement of #17623.

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by leng-yue** [comment]
Any update?

File: src/lightning/fabric/strategies/fsdp.py [hunk 1]
```diff
@@ -506,20 +507,37 @@ def load_checkpoint(
...
+            for optim_key, optim in optimizers.items():
+                # rank0_only should be false because we need to load the optimizer state on all ranks
+                with _get_full_state_dict_context(module, rank0_only=False):
+                    temp_state_dict = checkpoint[optim_key]
+
+                    # Handling the case where the optimizer state is saved from a normal optimizer
+                    if isinstance(list(checkpoint[optim_key]["state"].keys())[0], int):
+                        temp_state_dict = FSDP.rekey_optim_state_dict(
+                            temp_state_dict, OptimStateKeyType.PARAM_NAME, module
+                        )
```
Comment (awaelchli) [inline src/lightning/fabric/strategies/fsdp.py:526]: Any chance you could help me understand this a bit better? I've looked at this function `FSDP.rekey_optim_state_dict` in the past but never understood why this is needed. What do we consider a "normal optimizer" here. When we're loading from a full-state single file, isn't the state already "normal"?

File: tests/tests_fabric/strategies/test_fsdp_integration.py [hunk 2]
```diff
@@ -161,6 +166,34 @@ def test_fsdp_save_full_state_dict(tmp_path):
...
+
+    # verify the full state can be loaded back into a FSDP model/strategy
+    fabric = BoringFabric(
+        accelerator="cuda",
+        strategy=FSDPStrategy(auto_wrap_policy=always_wrap_policy, state_dict_type="full"),
+        devices=2,
+    )
+    fabric.run()
+    metadata = fabric.load(checkpoint_path, {"model": fabric.model, "optimizer": fabric.optimizer})
+    assert metadata == {"steps": 1}
```
Comment (awaelchli) [inline tests/tests_fabric/strategies/test_fsdp_integration.py:180]: I think this block is redundant since we are doing the same above. Could we combine the asserts for the optimizer with the code lines above? Btw, for loading, passing `state_dict_type="full"` is not necessary.

**Review by awaelchli — COMMENTED** [review]
@leng-yue That's awesome, thank you so much for jumping on this so quickly! I have minor comments. Great work and thanks for extending the test!

File: src/lightning/fabric/strategies/fsdp.py (hunk 1 above)
Comment (leng-yue) [reply src/lightning/fabric/strategies/fsdp.py:526]: When saving from a "normal" optimizer, the state dictionary is indexed by `id` instead of `param_name`. However, for some reason (possibly due to shards where each device contains a different portion of the state), the `optim_state_dict_to_load` expects the state dictionary to be indexed by `param_name`.

Some reference [here](https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel.rekey_optim_state_dict):
> Re-keys the optimizer state dict optim_state_dict to use the key type optim_state_key_type. This can be used to achieve compatibility between optimizer state dicts from models with FSDP instances and ones without.

Comment (leng-yue) [reply src/lightning/fabric/strategies/fsdp.py:526]: When saving from a "normal" optimizer, the state dictionary is indexed by `id` instead of `param_name`. However, for some reason (possibly due to shards where each device contains a different portion of the state), the `optim_state_dict_to_load` expects the state dictionary to be indexed by `param_name`.

Some reference [here](https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel.rekey_optim_state_dict):
> Re-keys the optimizer state dict optim_state_dict to use the key type optim_state_key_type. This can be used to achieve compatibility between optimizer state dicts from models with FSDP instances and ones without.

File: tests/tests_fabric/strategies/test_fsdp_integration.py (hunk 2 above)
Comment (leng-yue) [reply tests/tests_fabric/strategies/test_fsdp_integration.py]: I agree with your suggestion that using `state_dict_type="full"` is unnecessary. 

The purpose of the first mentioned block is to ensure that a model can load states saved from FSDP without FSDP (such as DDP). On the other hand, the second-mentioned block is responsible for restoring FSDP's weights. 

In the `update better unit test` commit, I added another test, so now we have:
- Save from FSDP, restore to FSDP
- Save from FSDP, restore to other strategy
- Save from other strategy, restore to FSDP

File: src/lightning/fabric/strategies/fsdp.py (hunk 1 above)
Comment (awaelchli) [reply src/lightning/fabric/strategies/fsdp.py:526]: Got it, thanks!

File: src/lightning/fabric/strategies/fsdp.py [hunk 3]
```diff
@@ -506,20 +507,37 @@ def load_checkpoint(
...
+
+                    optim_state_dict = FSDP.optim_state_dict_to_load(
+                        optim_state_dict=temp_state_dict,
+                        model=module,
+                        optim=optim,
+                    )
+                    optim.load_state_dict(optim_state_dict)
+
+                # optim.load_state_dict(optim_state_dict, strict=True)
```
Comment (awaelchli) [inline src/lightning/fabric/strategies/fsdp.py:535]:
```suggestion
```

File: src/lightning/fabric/strategies/fsdp.py [hunk 4]
```diff
@@ -506,20 +507,37 @@ def load_checkpoint(
...
             # There is currently no other way because `summon_full_params` does not support write-back from rank 0 only.
             checkpoint = torch.load(path, map_location="cpu")
             with FSDP.summon_full_params(module, writeback=True, rank0_only=False):
                 module.load_state_dict(checkpoint.pop(module_key), strict=strict)
 
+            # Load optimizer states
+            for optim_key, optim in optimizers.items():
+                # rank0_only should be false because we need to load the optimizer state on all ranks
+                with _get_full_state_dict_context(module, rank0_only=False):
+                    temp_state_dict = checkpoint[optim_key]
```
Comment (awaelchli) [inline src/lightning/fabric/strategies/fsdp.py:520]:
```suggestion
                    temp_state_dict = checkpoint.pop(optim_key)
```

**Review by awaelchli — APPROVED** [review]

**Comment by leng-yue** [comment]
I don't know why some tasks failed after I changed few lines, and there is no error message in azure pipelines...
